File: lunarrx-core/src/main/java/com/cisco/vss/lunar/rx/plugin/core/LunarTrackItemStreamGenerator.java
package com.cisco.vss.lunar.rx.plugin.core;

import rx.Observable;
import static com.cisco.vss.lunar.rx.plugin.core.LunarConversions.*;
import rx.functions.Func2;
import rx.functions.Func1;

public class LunarTrackItemStreamGenerator<R extends LunarTrackItem> implements Func2<Observable<byte[]>, LunarTrack, Observable<? extends byte[]>> {
	protected final Class<R>                                           resultType;
	private   final Func1<Observable<byte[]>, Observable<? extends R>> transform;

	protected LunarTrackItemStreamGenerator(final Class<R> resultType, final Func1<Observable<byte[]>, Observable<? extends R>> transform) {
		this.resultType = resultType;
		this.transform  = transform;
	}

	@Override
	public Observable<byte[]> call(final Observable<byte[]> input, final LunarTrack resultTrack) {
		return this.transform.call(input)
			   .map(setTrackDetails(resultTrack, resultType))
			   .map(object2JsonString(resultType))
			   .map(string2Byte);
	}
}

File: lunarrx-scala/src/main/scala/com/cisco/vss/lunar/rx/plugin/scala/LunarTrackItemStreamGenerator.scala
package com.cisco.vss.lunar.rx.plugin.scala
import com.cisco.vss.lunar.rx.plugin._
import rx.lang.scala.JavaConversions._
import rx.lang.scala._
import com.cisco.vss.lunar.rx.plugin.core.TrackItem

abstract class LunarTrackItemStreamGenerator[R <: TrackItem] (lunar: Lunar, sourceTrackTemplate: core.LunarTrack, resultType: Class[R]) {
	private [scala] val asJava: core.LunarTrackItemStreamGenerator[R] = new core.LunarTrackItemStreamGenerator[R](
	    lunar.asJavaLunar, 
	    sourceTrackTemplate,
	    resultType
	){
      @Override
      def generateR(input: rx.Observable[Array[Byte]]) : rx.Observable[_ <: R] = {
        val result     = transformA(toScalaObservable(input))
        val javaResult = toJavaObservable(result) 
        return javaResult
      }
	}
	
    def run: Unit = asJava.run()

		//TODO: name!???
	def transformA(input: Observable[Array[Byte]]): Observable[R]

}